### ISSUE TITLE
CCurlFile::Stat  Failed: Server returned nothing (no headers, no data)

### ISSUE THREAD

**Issue by khinch** [petegallagher/plugin.video.zoneminder#2]
Hi, 
Just installed this and I'm getting the following when I try to connect to a camera:

11:55:55.932 T:140030764557504   ERROR: CCurlFile::Stat - Failed: Server returned nothing (no headers, no data)(52) for http://<redacted>/zm/cgi-bin/nph-zms?scale=auto&width=1920&height=1080&mode=jpeg&maxfps=25&monitor=2&user=&pass=
11:55:55.942 T:140030764557504  NOTICE: VideoPlayer: Opening: http://<redacted>/zm/cgi-bin/nph-zms?scale=auto&width=1920&height=1080&mode=jpeg&maxfps=25&monitor=2&user=&pass=
11:55:55.942 T:140030764557504 WARNING: CDVDMessageQueue(player)::Put MSGQ_NOT_INITIALIZED
11:55:56.484 T:140030764557504   ERROR: CCurlFile::Stat - Failed: Server returned nothing (no headers, no data)(52) for http://<redacted>/zm/cgi-bin/nph-zms?scale=auto&width=1920&height=1080&mode=jpeg&maxfps=25&monitor=2&user=&pass=
11:55:56.484 T:140027862439680  NOTICE: Creating InputStream
11:55:56.739 T:140027862439680   ERROR: CCurlFile::FillBuffer - Failed: Server returned nothing (no headers, no data)(52)
11:55:56.739 T:140027862439680   ERROR: CCurlFile::Open failed with code 0 for http://<redacted>/zm/cgi-bin/nph-zms?scale=auto&width=1920&height=1080&mode=jpeg&maxfps=25&monitor=2&user=&pass=
11:55:56.968 T:140027862439680   ERROR: CCurlFile::FillBuffer - Failed: Server returned nothing (no headers, no data)(52)
11:55:56.968 T:140027862439680   ERROR: CCurlFile::Open failed with code 0 for http://<redacted>/zm/cgi-bin/nph-zms?scale=auto&width=1920&height=1080&mode=jpeg&maxfps=25&monitor=2&user=&pass=
11:55:56.968 T:140027862439680   ERROR: Open - failed to open source <http://<redacted>/zm/cgi-bin/nph-zms?scale=auto&width=1920&height=1080&mode=jpeg&maxfps=25&monitor=2&user=&pass=>
11:55:56.968 T:140027862439680   ERROR: CVideoPlayer::OpenInputStream - error opening [http://<redacted>/zm/cgi-bin/nph-zms?scale=auto&width=1920&height=1080&mode=jpeg&maxfps=25&monitor=2&user=&pass=]
11:55:56.968 T:140027862439680  NOTICE: CVideoPlayer::OnExit()
11:55:56.969 T:140030764557504   ERROR: Playlist Player: skipping unplayable item: 0, path [plugin://plugin.video.zoneminder/?action=play&video=http%3A%2F%2F<redacted>%3A<redacted>%2Fzm%2Fcgi-bin%2Fnph-zms%3Fscale%3Dauto%26width%3D1920%26height%3D1080%26mode%3Djpeg%26maxfps%3D25%26monitor%3D2%26user%3D%26pass%3D]
11:55:56.973 T:140030764557504  NOTICE: CVideoPlayer::CloseFile()
11:55:56.973 T:140030764557504  NOTICE: VideoPlayer: waiting for threads to exit
11:55:56.973 T:140030764557504  NOTICE: VideoPlayer: finished waiting
11:55:56.978 T:140030764557504  NOTICE: CVideoPlayer::CloseFile()
11:55:56.978 T:140030764557504  NOTICE: VideoPlayer: waiting for threads to exit
11:55:56.978 T:140030764557504  NOTICE: VideoPlayer: finished waiting

This is with ZoneMinder 1.32.3 and Kodi 17.6 (Debian). User has view permissions to everything. Any suggestions much appreciated.

**Comment by petegallagher** [comment]
Hi your error message is showing that the Zoneminder base URL is not configured correctly i.e.:

```
http:///zm/cgi-bin/
```

**Comment by khinch** [comment]
Interesting, github has removed data there. I changed my IP and port to "redacted" inside angled brackets. Github must have noticed that up and removed it entirely! 

When I open the addon it lists my camera with the correct name so it must be connecting to ZM successfully, it's only when I try to connect to the feed that I see the error above. 

I've also noticed that if I copy & paste this into a browser then I see the feed:

http://{realip}:{realport}/zm/cgi-bin/nph-zms?scale=auto&width=1920&height=1080&mode=jpeg&maxfps=25&monitor=2&user={realuser}&pass={realpassword}

**Comment by petegallagher** [comment]
1.  So I assume your log message above has also had the values for `&user=username&pass=password` redacted? I.e. are you able to copy/paste the URL directly from your log file and retrieve the stream in your browser or via `curl`?
2. Are there any error messages on Zoneminder when this fails in Kodi?
3. What are you Zoneminder auth settings e.g. `OPT_USE_AUTH`, `AUTH_TYPE`, `AUTH_RELAY`, `AUTH_HASH_IPS` and `AUTH_HASH_LOGINS`?

**Comment by khinch** [comment]
I think we can mark this one as solved. Changing AUTH_RELAY from "none" to "hashed" solved this and I can now view the feed. I am using QuantumObject's docker image for ZoneMinder and the docs said if AUTH_RELAY was anything other than "none" then live feeds would be unavailable, but I think this is outdated as it's working fine for me now. 

To answer you questions though, just in case the info helps anyone else:
1 - Yes, the log message had the values for user= and pass= redacted in the query params. I didn't redact them though, they were like that in the logs, so Kodi must've done it. When I pasted a complete URL with correct credentials into Firefox I saw the feed. 
2 - I didn't check actually. It's running in a detached docker container. I need to figure out how to attach to it with a terminal at some point. 
3 - OPT_USE_AUTH=true, AUTH_TYPE=builtin, AUTH_RELAY=hashed (was none when the error occurred), AUTH_HASH_IPS=true,  AUTH_HASH_LOGINS=false

So thanks for all your help, and also thanks for the plugin!

**Comment by petegallagher** [comment]
> I am using QuantumObject's docker image for ZoneMinder

FYI there are official [Zoneminder containers available on Docker Hub](https://hub.docker.com/r/zoneminderhq/zoneminder). I've created the Ubuntu one, example usage:

```
docker run -d -t -p 1080:80 \
    -e TZ='Europe/London' \
    -v ~/zoneminder/events:/var/cache/zoneminder/events \
    -v ~/zoneminder/images:/var/cache/zoneminder/images \
    -v ~/zoneminder/mysql:/var/lib/mysql \
    -v ~/zoneminder/logs:/var/log/zm \
    --shm-size="512m" \
    --name zoneminder \
    zoneminderhq/zoneminder:latest-ubuntu18.04
```

> Yes, the log message had the values for user= and pass= redacted in the query params. I didn't redact them though, they were like that in the logs, so Kodi must've done it. When I pasted a complete URL with correct credentials into Firefox I saw the feed.

This was most likely because `AUTH_RELAY=none`. Kodi wouldn't have redacted this.

> I need to figure out how to attach to it with a terminal at some point.

Assuming you've given it a friendly name like in the example above you can use:

```
docker exec -it zoneminder bash
```

Otherwise use `docker ps` to get the current name.

**Comment by khinch** [comment]
Thanks! Got connected to a terminal inside the container, so that's good to know. Tried running the zoneminderhq container and I got connected in my browser but it wouldn't connect to my camera, with the logs reporting permission errors for some of the directories. I didn't have a chance to fully look into it, so I'll try again later and raise any issues there if I'm still having them. It's probably my config.
